### PR TITLE
Project 2: loop invariant code motion and loop reduction

### LINKED ISSUE #52: Project 2 Proposal: Strength reduction with loop invariant code motion
**What will you do?**
We will implement a strength reduction compiler optimization with loop invariant code motion and induction variable analysis. The goal is to replace the expensive operations with equivalent and cheaper operations (i.e., less execution time/power/area).

**How will you do it?**
We will first go through the code to identify all the loops and recognize the characteristics of local values within those loop. Here, we want to find out:
    * loop invariants (values not changed within the loop body)
    * induction variables (values iterated each time through the loop)
Now, as we have identified all the loop invariants and induction variables, we can hoist the loop invariants outside the body and replace some expressions inside the loop with the constants. After that, substitute expensive expressions composed of induction variables with cheaper operations (i.e., multiplication -> addition).

**How will you empirically measure success?**
As we are trying to reduce expensive operations to cheaper ones, so the performance can be measured as "the total number of arithmetic operations of each kind" (e.g., #addition, #multiplication) before and after code optimization. Also, we can run the test programs before and after code optimization to derive a wall clock speedup.

**Team members:**
 @tissue3 @neiladit

### PR DESCRIPTION
**Pull request by tissue3**
Closes #52

### CODE REVIEW

File: content/blog/2019-10-30-loop-reduction/index.md [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+
+
+
+## Evaluation
+
+1. We would like to evaluate our optimizer in two aspects: theoretical improvement and actual wall-clock speedup.
+
+   ### Benchmarks
+
+   Because loop invariant code motion and strength reduction are specific optimization for loops, and studies are based on case by case, we can hardly generate automatic testing scripts while writing bril codes to generate `for` loop are actually painful. Instead we reply on Typescript for help.
```
Comment (sampsyo) [inline content/blog/2019-10-30-loop-reduction/index.md]: - "studies are based on case by case": I'm not entirely sure what this means?
- capitalize "Bril"
- "writing Bril codes to generate `for` loop": But Bril code doesn't *generate* `for` loops… rather, `for` loops in higher-level languages generate Bril code!
- reply -> rely
- capitalize "TypeScript"

File: content/blog/2019-10-30-loop-reduction/index.md [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+## Evaluation
+
+1. We would like to evaluate our optimizer in two aspects: theoretical improvement and actual wall-clock speedup.
+
+   ### Benchmarks
+
+   Because loop invariant code motion and strength reduction are specific optimization for loops, and studies are based on case by case, we can hardly generate automatic testing scripts while writing bril codes to generate `for` loop are actually painful. Instead we reply on Typescript for help. 
+
+   However, default Bril scripts generated by Typescript always introduces redundancy during translation, and code motion optimization by default will be performed. For example, `int v10 = const 5; int v11 = v11 + v10;` can be found in the body of a loop if there is a line `x = x + 5` in Typescript. Because we don't perform other optimizations, we wonder how much speedup is contributed by the redundancy. We therefore write test cases with no expected optimizations (but will have some when translated to Bril) as baseline to compare actual wall-clock speedup.
```
Comment (sampsyo) [inline content/blog/2019-10-30-loop-reduction/index.md]: Again, capitalize TypeScript. (Apply throughout.)

Comment (sampsyo) [inline content/blog/2019-10-30-loop-reduction/index.md]: `int v10 = const 5` is not Bril syntax. (Types come after the variables.)

File: content/blog/2019-10-30-loop-reduction/index.md [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+   | `normal.ts`, `nested.ts`, `nestedif.ts` | No optimization expected.                   |
+   | `codemotion*.ts`                        | Expecting code motion.                      |
+   | `strengthreduction*.ts`                 | Expecting strength reduction optimizations. |
+   | `both*.ts`                              | Expecting both kinds of optimizations.      |
+
+
+
+   ### Implicit Redundancy
+
+   Bril scripts generated by Typescript introduce some redundant operations usually an `id` or a `const`. Inside a loop, these would usually be loop invariant code which gets removed by our optimizer. However, we were not sure if this would contribute towards outr speedup. To make sure that these are actually negligible we benchmarked similar codes without explicit loop invariant codes in the loop to see speedups (if any).
```
Comment (sampsyo) [inline content/blog/2019-10-30-loop-reduction/index.md]: "some redundant operations: usually" (insert some sort of punctuation; probably a colon)
"outr" (please run spell check)

File: content/blog/2019-10-30-loop-reduction/index.md [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+   | `codemotion2.ts`        | 908.6 ms            | 665.4 ms          | 1.36x   |
+   | `strengthreduction1.ts` | 192.1 ms            | 173.2 ms          | 1.11x   |
+   | `strengthreduction2.ts` | 61.1 ms             | 56.3 ms           | 1.09x   |
+   | `both1.ts`              | 655.0 ms            | 484.1 ms          | 1.35x   |
+   | `both2.ts`              | 66.2 ms             | 49.8 ms           | 1.33x   |
+
+   Based on the evaluation results, we find out that both loop optimization techniques can provide some speedup. Loop invariant code motion can provide up to 1.36x speedup, while strength reduction can only have about 10% performance improvement. That is due to the fact that multiplication/division does not take significantly more cycles than addition/substraction on modern machines and there are many redundant `id` operations when generating `bril` programs from `ts`. Thus, only limited performance improvement can be obtained We expect that more speedup can be obtained if exponential operation is optimized. Also, note that there are nested loops in `codemotion2.ts` and `both1.ts`, thus their runtime are much longer than the rest.
```
Comment (sampsyo) [inline content/blog/2019-10-30-loop-reduction/index.md]: This is a good, detailed evaluation! Nice work. You might have gotten a little more detail about the performance by tracing the counts of instruction types—for example, showing that you do decrease the number of multiplication instructions executed could be cool, even if the number add instructions goes up.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Great work! This was a fun read. I have several writing comments within.

File: content/blog/2019-10-30-loop-reduction/index.md [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+After that, we just need to find all exit blocks where `d` is live-out and check if `d`'s' block dominates all loop exits.
+
+```python
+edest = [e for e in exits[back_edge] if instr['dest'] in live_var[1][e]]
+all([b_name in dom[e] for e in edest]) # if true, third check passed.
+```
+
+### Block to Dictionaries
+
+` json.load(sys.stdin)['functions']` gives us dictionaries and for each dictionary we obtain list of instructions when the key is `instrs`. Then we change this list of instructions into a directory of blocks. We would like to reverse this process to regenerate list of instructions with modified blocks. However, original block does not have so many labels introduced when generating block names and creating pre-headers. Luckily, the modified blocks are still ordered dictionary and we can omit blocks we introduced. Therefore in this step, we only create `label` instruction when the original labels in blocks. The rest is just copy every instruction other than labels to the new list of instructions.
```
[the remark this answers is not in the document]
Comment (tissue3) [reply content/blog/2019-10-30-loop-reduction/index.md]: For example, main function does not have a label, but to form a block, we generate a `b0` for its label. Similarly, when generating a pre-header, we might have inserted blocked labeled with `b1` .  Luckily, the modified blocks are still in an ordered dictionary. Blocks with no label at first is safe to be unlabled and for `pre-header` blocks, we assume they have only one in-edge and one out-edge.
So it is safe here to just delete `label` that are not originally there.

[the remark this answers is not in the document]
Comment (sampsyo) [reply content/blog/2019-10-30-loop-reduction/index.md]: Got it. It might be wort explaining this fact in your post.

For what it's worth, the reason to create pre-headers in the first place is typically that loop headers can have multiple predecessors. So pre-headers generally do too. The idea is to make it so the loop _header_ has only a single predecessor, which is a place you can put pre-loop code. I think you've just assumed that away, but I wanted to point out that this is nonstandard. If you require loop headers to have only a single predecessor, then there is no real reason to create pre-headers—you can just put the code right in that predecessor.

[the remark this answers is not in the document]
Comment (tissue3) [reply content/blog/2019-10-30-loop-reduction/index.md]: Yes, I've put the explanation in the updated post. Thanks for pointing out that the previous explantion is hard to follow.
We decided to create pre-header blocks because we work on LICM and LR separately, and inserting instructions to an empty block might result in less error when we try to merging things together. And yes, the pre-header is not compulsory rather than just a design choice. I will put this line in the post either.

File: content/blog/2019-10-30-loop-reduction/index.md [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+```
+
+
+
+### Reaching Definition
+
+For finding loop invariant code, we look into reaching definitions of the argument and check wether they are outside the loop. The reaching definition problem was discussed in the lecture of CS 6120. For each block we can generate a list of reaching inputs and outputs. The data structure that we chose to represent this is a dictionary of dictionaries. We decided to store the block number associated with the variable in each block. Hence reaching definitions of each block has a dictionary of variables (as keys) and block number they were defined in (as values). This helps in keeping track of reaching definitions outside the loop to identify loop invariant code.  The reaching definitions look something like this:
+
+```
+reaching_defs = {blocks:{variable:[list of block numbers]}}
```
[the remark this answers is not in the document]
Comment (neiladit) [reply content/blog/2019-10-30-loop-reduction/index.md]: Excellent observation! I do _assume_ that it is defined once in each block while I'm storing it in the list since I just need to know if the reaching definition is outside the loop or not. Does that make sense?

[the remark this answers is not in the document]
Comment (sampsyo) [reply content/blog/2019-10-30-loop-reduction/index.md]: Sure; I think that seems safe!

**Review by sampsyo — CHANGES_REQUESTED** [review]
Hi, folks—this is looking good, but I just noticed that the (metadata) title is wrong. Can you please pick a title that describes your project?

Also, the second half of your post is still indented 3 spaces for some reason, although this seems not to cause any rendering glitches in Zola, as I expected it might have…

**Comment by sampsyo** [comment]
Sounds good; thanks!
